---
title: 'MetaboDiff tutorial'
author: | 
  | Andreas Mock
  | Experimental Neurosurgery, Heidelberg University Hospital,
  | Division of Applied Bioinformatics, German Cancer Research Center (DFKZ) Heidelberg &
  | Department of Medical Oncology, National Center for Tumor Diseases (NCT) Heidelberg
date: '`r Sys.Date()`'
output: BiocStyle::html_document
fig_caption: yes
vignette: >
  %\VignetteIndexEntry{Cancer Variant Explorer (CVE) tutorial}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{utf8}
---

```{r setup, include=FALSE}
library("devtools")
library("gdata")
library("SummarizedExperiment")
install_github("andreasmock/MetaboDiff")
```

# Introduction

Comparative metabolomic data analysis comes of age. Although access to a combined ultrahigh performance liquid chromatography/tandem mass spectrometry (LC/MC) and gas chromatography/mass spectrometry (GC/MS) pipeline is still limited, an increasing list of commercial vendors offer reproducible high-quality metabolomic data. 

Unlike high-throughput technologies assessing other molecular layers of biology (e.g. sequencing or microarrays for genomics and transcriptomics), open-source computational tools for comparative metabolomics are scarce.

The `MetaboDiff` package enables open-source comparative metabolomic analysis.

# Installation

`MetaboDiff` can be installed via Bioconductor

```{r,eval=FALSE}
source('http://www.bioconductor.org/biocLite.R')
biocLite('MetaboDiff')
```

and once installed loaded by

```{r,eval=FALSE, message=FALSE}
library("MetaboDiff")
```

# Reading data and annotation

## Input files
To ensure a seamless analysis, the metabolomic data is stored as an `SummarizedExperiment` object [^1]. Hence, the following input object are needed 

- `assay` - a matrix containing the relative metabolomic measurements with sample IDs as the colnames and metabolite IDs as the rownames (HMDB, KEGG or ChEBI IDs). 
the Human Metabolome Database (HMDB) IDs as the rownames.
- `colData` - a dataframe containing the sample metadata.
- `rowData` - a dataframe containing the metabolite annotation.

## Example data

```{r, eval=TRUE, message=FALSE}
# read xlsx sheet 
data = as.matrix(read.xls("http://cancerres.aacrjournals.org/highwire/filestream/291231/field_highwire_adjunct_files/2/132870_2_supp_2886796_nkp7r4.xlsx",sheet=2))
```

### Creation of assay object containing metabolomic data

```{r, eval=TRUE}
assay = apply(data[6:nrow(data),6:ncol(data)],2,as.numeric)
# patient IDs as colnames
colnames(assay) = as.vector(data[1,6:ncol(data)])
# HMDB-IDs as rownames
rownames(assay) = data[6:nrow(data),5]
# remove metabolites with missing HMDB-IDs
assay = assay[!rownames(assay)=="",]
# display head of assay
assay[1:4,1:5]
```

### Creation of colData object containing sample metadata

```{r}
# create colData
colData = as.data.frame(t(data[1:5,6:ncol(data)]))
names(colData) = data[1:5,5]
rownames(colData) = NULL
# show input_matrix object
head(colData)
```

### Creation of rowData object contaning metabolite annotation

`MetaboDiff` supports all common metabolic IDs as input for the annotation (HMDB, KEGG and ChEBI). The function `get_SMPDBanno` retrieves annotation from the Small Molecule Pathway Database (SMPDB)[^2]. The current build used within `MetaboDiff` is version 2.0 and will be updated as new versions are released. Please refer to the development branch of `MetaboDiff` to work with the latest SMPDB annotation.

```{r}
rowData = get_SMPDBanno(rownames(assay))
rownames(rowData) = rownames(assay)
missing_anno = which(is.na(rowData$SMPDB.ID))
```

## Create SummarizedExperiment object

```{r}
met = SummarizedExperiment(assays=assay[-missing_anno,],
                           rowData=rowData[-missing_anno,],
                           colData=colData)
met
```

# Data preprocessing and quality control

# Unsupervised analysis

# Compartative analysis

# Visualization of metabolic networks

To compare the structure of metabolic networks across tumor entities, network visualizations should ideally be perceptual uniform. However, networks are usually visualized using force-directed layouts (e.g. Fruchterman Reingold[^2]) generating pleasant but hardly interpretable networks earning the nickname ’hairballs’.

## Hive plots

To address this issue, Kryzwinski et al. have developed an algorithm in which nodes are placed on radially oriented linear axes according to a well-defined coordinate system[^3]. In resemblance to the top of a beehive, the authors termed these visualizations hive plots and show that they can be adjusted to suit the purpose, are easily explained and understood and most importantly can generate useful and quantitatively inter- pretable results.

# Differentially abundant subpathways

graph clustering - hotnet2 

# Session information

```{r}
sessionInfo()
```

# References
[^1]: M. Morgan, V. Obenchain, J. Hester, and P. Herv, “Summarizedexperiment: Summarized- experiment container,” R package version, vol. 1.2.3, 2016.

[^2]: Jewison T, Su Y, Disfany FM, et al. SMPDB 2.0: Big Improvements to the Small Molecule Pathway Database Nucleic Acids Res. 2014 Jan;42(Database issue):D478-84.

[^3]: T. M. J. Fruchterman and E. M. Reingold, “Graph drawing by force-directed placement,” Software: Practice and Experience, vol. 21, no. 11, pp. 1129–1164, 1991.

[^4]: M. Krzywinski, I. Birol, S. J. M. Jones, and M. A. Marra, “Hive plots–rational approach to visualizing networks.,” Briefings in Bioinformatics, vol. 13, pp. 627–644, Sept. 2012.


