---
title: "MetaboDiff tutorial"
author:
- name: Andreas Mock
  affiliation: University Hospital and National Center for Tumor Diseases (NCT), Heidelberg
  email: andreas.mock@med.uni-heidelberg.de
package: MetaboDiff
output:
  BiocStyle::html_document:
    toc: true
    toc_float: true
abstract: |
    Comparative non-targeted metabolomics comes of age through commercial vendors offering metabolomics for translational researchers outside the mass spectrometry field. The MetaboDiff packages aims to provide a low-level entry to differential metabolomic analysis by starting off with the table of relative metabolite measurements. In addition, it introduces the usage of weighted correlation network analysis for metabolomics.
vignette: |
  %\VignetteIndexEntry{MetaboDiff}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}  
---

# Getting started

The `MetaboDiff` R package requires R version $\geq$ 3.4. 

## Installing dependencies

CRAN occasionally fails to compile the `WGCNA` package for Mac OS X. Hence, it is recommended to install the package before installing `MetaboDiff`:

```{r dependencies, eval=FALSE}
install.packages("WGCNA")
```

If asked, install the package from source. Alternatively you might need to download the package from the [developer](https://labs.genetics.ucla.edu/horvath/CoexpressionNetwork/Rpackages/WGCNA/) and install the package locally:

```{r local download, eval=FALSE}
install.packages(path_to_file, repos = NULL, type="source")
```

If you encounter problems installing `WGCNA`, please refer to the [developer page](https://labs.genetics.ucla.edu/horvath/CoexpressionNetwork/Rpackages/WGCNA/). Please note that `MetaboDiff` can only be installed if `WGCNA` was successfully installed.

## Installing MetaboDiff

`MetaboDiff` can be installed via Github

```{r installation, eval=FALSE}
library("devtools")
install_github("andreasmock/MetaboDiff")
```

and once installed loaded by

```{r loading, message=FALSE}
library("MetaboDiff")
```

# Part I: Data processing

## Input data

MetaboDiff requires three objects as input:

* `assay` - a matrix containing the relative metabolic measurements
* `rowData` - a dataframe containing the available metabolite annotation
* `colData` - a dataframe containing sample metadata 

Example data for the tutorial is derived from a study by Priolo and colleagues in which the authors used the service of Metabolon&reg; to compare the tissue metabolome of 40 prostate cancers with 16 normal prostate specimens[^1].

Please make sure that the colnames (samples) of `assay` correspond to the rownames (patients) of `colData` and the rownames of `assay` to the rownames of `rowData`. 

```{r}
assay[1:5,1:5]
head(colData)
head(rowData)
```

The function `create_mae` merges all objects into a so called `MultiAssayExperiment` object[^2] to simplify all downstream analysis.

```{r, warning=FALSE}
(met <- create_mae(assay,rowData,colData))
```

## Metabolite annotation

Metabolite annotation can be retrieved from the Small Molecular Pathway Database (SMPDB) if HMDB, KEGG or ChEBI ids are part of the rowData object [^3].

```{r}
# Please specify the accoding column in rowData. In the example data set, the KEGG id was in column 6 and the HMBD id in column 7. A ChEBI id was not available.
met <- get_SMPDBanno(met,
                        column_kegg_id=6,
                        column_hmdb_id=7,
                        column_chebi_id=NA)
```

## Imputation of missing values

In contrast to other high-throughput technologies, missing values are common in quantitative metabolomic datasets. 

The function `na_heatmap` visualizes the missing metabolite measurements across the samples. The name of the column in colData for grouping and the label colors for the two groups need to be specified.

```{r heat, fig.cap="Missing metabolic measurements across the example data set. Missing measurements are visualized by a binary heatmap and barplots summarizing the fraction of missing measurement per metabolite and sample, respectively. In addition, the group label of interest (tumor (T) vs. normal (N)) is visualized. By default, metabolite measurements are only imputed if less than 40 percent are missing."}
na_heatmap(met,
           group_factor="tumor_normal",
           label_colors=c("darkseagreen","dodgerblue"))
```

The example data supports the need for data imputation (figure \@ref(fig:heat)). It could be shown that k-nearest neighbor imputation minimizes the effects on the normality and variance of the data as long as the number of missing data does not exceed 40%[^4]. 

The function `knn_impute` adds the slot "impute" to the MultiAssayExperiment object that contains the imputed relative metabolite measurements for all metabolites with raw measurements in more than 60% of cases. We recommend a cutoff of 40% (i.e. 0.4). However the cutoff might be changed according to the discretion of the user.

```{r imputation}
(met = knn_impute(met,cutoff=0.4))
```

As apparent form the summary description of the object `met` the imputed data matrix contains only 238 of the 307 original metabolites according to the cutoff of 40% missing values.

## Removal of outliers

Before we normalize the data, we want to exclude outliers in the study set. To this end, the function `outlier_heatmap` is provided.
The sample annotation shows the number of missing metabolites per sample as a proxy of the impact of imputation on clustering. 
To showcase outliers, the hierarchical clustering tree is cut in 2 clusters (figure \@ref(fig:outlier)).

```{r outlier, fig.cap="Hierarchical clustering of metabolite measurements. The tree is cut in 2 clusters. The user has the choice to exclude one cluster which he thinks might represent outliers. To determine the effect of imputation, a barplot displaying the fraction of missing metabolites is shown in the column annotation of the heatmap."}
outlier_heatmap(met,
                group_factor="tumor_normal",
                label_colors=c("darkseagreen","dodgerblue"))
```

The imputed data of the example study set displays a cluster of 5 samples (cluster 1) with in average lower relative metabolite measurements. Due to the lack of batch information, this cannot be investigated further at this time. To demonstrate how a cluster can be removed, we apply the function `remove_cluster` to remove cluster 1:

```{r remove_cluster, message=FALSE}
(met <- remove_cluster(met,cluster=1))
```

As displayed in the summary of the `met` object, the 5 samples of cluster 1 were successfully removed from the slots "raw" and "imputed".

## Normalization

Variance stabilizing normalization (vsn) is used to ensure that the variance remains nearly constant over the measured spectrum[^5].

```{r norm, message=FALSE}
(met <- normalize_met(met))
```

At this point the data processing is completed with the `MultiAssayExperiment` object containing 4 slots:

- raw - raw relative metabolic measurements as provided by company or core facility
- imputed - imputed relative metabolic measurements (k-nearest neighbor imputation)
- norm - normalized relative metabolic measurements (vsn)
- norm_imputed - normalized and imputed relative metabolic measurements (vsn)

## Quality control of normalization

```{r quality, message=FALSE, warning=FALSE, fig.cap="Quality control plot. Boxplot displaying the distribution of (A) raw, (B) imputed (C) normalized and (D) imputed and normalized relative metabolite measurements for all samples of the study set. Boxplots are colored according to the grouping of interest, i.e. tumor vs. normal.", fig.width=10, fig.height=8}
quality_plot(met,
             group_factor="tumor_normal",
             label_colors=c("darkseagreen","dodgerblue"))
```

The quality control plots shows the distribution of raw and normalized metabolic measurements for every sample in the study set. As aimed, the distribution of measurements is comparable across the study set for the normalized and imputed data (figure \@ref(fig:quality)D).

# Part II: Data analysis

Part II of `MetaboDiff` comprises a set of functions for comparative data analysis that are purpose-built for the preprocessed `MultiAssayExperiment` object. All statistics obtained in this part will be saved in the metadata slot of the object. Certain plotting functions will only be available once the corresponding statistics has been run, e.g. you will only be able to plot the volcano plot after executing the function for differential analysis (`diff_test`). Please note that the data analysis has been developed for metabolomic data sets of more than 100 metabolites. Hence, some functionality might not be available with smaller data sets < 50 metabolites.

## Biological questions

The following table summarizes the biological questions that can be answered by MetaboDiff functions.

Question | Statistical Methodology
----------------------------------------- | ----------------------------------
Are there metabolome-wide changes between samples? | unsupervised analyses (PCA, tSNE)
Which individual metabolites show differential abundance between groups? | supervised analyses = hypothesis testing with correction for multiple testing
Which metabolic subpathways differ between groups? | weighted correlation network analysis


## Unsupervised analysis

To explore the metabolomic profiles between samples in an unsupervised fashion, MetaboDiff offers PCA as a linear and tSNE as a non-linear dimension reduction technique.

### PCA 

```{r pca, message=FALSE, fig.cap="PCA plot. Samples are colored according to the group factor."}
pca_plot(met,
         group_factor="tumor_normal",
             label_colors=c("darkseagreen","dodgerblue"))
```

A plot of the first two principal components (figure \@ref(fig:pca)) shows no metabolome-wide difference between the grouping of interest.

### tSNE 


```{r tSNE, message=FALSE, fig.cap="tSNE plot. Samples are colored according to the group factor."}
tsne_plot(met,
          group_factor="tumor_normal",
             label_colors=c("darkseagreen","dodgerblue"))
```

The first two dimensions (figure \@ref(fig:tSNE)) in the tSNE plot do not reveal a distinct difference in the metabolomic profiles between the normal and tumor samples.

## Hypothesis testing

Differential analysis for individual metabolites is performed using Student's T-Test. Correction for multiple testing is performed by Benjamini-Hochberg procedure. 

Hypothesis testing will be performed for the sample grouping "tumor_normal", as well as for the randomly generated grouping "random_gender".

```{r diff_test, message=FALSE}
met = diff_test(met,
                group_factors = c("tumor_normal","random_gender"))
```

The results of the hypothesis testing is saved in the metadata slot of the `MultiAssayExperiment` object. 

```{r str_met}
str(metadata(met), max.level=2)
```

The columns of the result dataframe are the unadjusted p-value (pval), the adjusted p-value by independent hypothesis weighting (adj_pval), the Fold-Change between groups (fold_change).

The comparative analysis can be visualized by means of a volcano plot (figure \@ref(fig:volcano1)).

```{r volcano1, fig.cap="Volcano plot of comparison tumor vs normal. The dashed lines mark the significance thresholds adjusted p-value<0.05 and fold-change>1.5. Significant metabolites are colored."}
volcano_plot(met, 
             group_factor="tumor_normal",
             label_colors=c("darkseagreen","dodgerblue"))
```

As a sanity check, we also display the volcano plot for the random grouping "random_gender" for which we do not expect the same number of significant metabolites.

```{r volcano2, fig.cap="Volcano plot of random comparison. The dashed lines mark the significance thresholds adjusted p-value<0.05 and fold-change>1.5."}
volcano_plot(met, 
             group_factor="random_gender",
             label_colors=c("brown","orange"))
```

As expected, no metabolite was significantly different between the randomly assigned grouping male vs. female after multiple testing with a cutoff of p-value<0.05 (figure \@ref(fig:volcano2)).

## Metabolic correlation network analysis 

### Implementation

To derive meaningful subpathways that are enriched between groups, MetaboDiff generates a metabolic correlation network. Within MetaboDiff, metabolic correlation network analysis is performed by a set of functions:

* `diss_matrix` - construction of dissimilarity matrix
* `identify_modules` - identification of metabolic correlation modules
* `name_modules` - name metabolic correlation modules
* `calculate_MS` - calculation of module signficance to relate sample traits to modules

For more details about the implementation and interpretation, please refer to the package vignette "Metabolic correlation network analysis".

```{r WGCNA}
met <- met %>%
    diss_matrix %>%
    identify_modules(min_module_size=5) %>%
    name_modules(pathway_annotation="SUB_PATHWAY") %>%
    calculate_MS(group_factors=c("tumor_normal","random_gender"))
```

### Association of sample traits with correlation modules

The following plots illustrate the association of the sample trait of interest (tumor vs. normal) with the metabolic correlation modules derived form the data.

```{r MSplot, fig.cap="Association of tumor vs normal trait with metabolic correlation modules. Module 2 (Creatine metabolism / Glutathione metabolism) showed a significant module significance."}
MS_plot(met,
        group_factor="tumor_normal",
        label_colors=c("darkseagreen","dodgerblue"),
        p_value_cutoff=0.05)
```

```{r MSplot2, fig.cap="Association of random trait with metabolic correlation modules. No significant module significance can be observed."}
MS_plot(met,
        group_factor="random_gender",
        label_colors=c("brown","orange"),
        p_value_cutoff=0.05)
```

In line with the volcano plots, the random grouping did not result in a significant association (figure \@ref(fig:MSplot2))

### Exploration of module of interest

The relationship of metabolites within a correlation module can be explored by means of a module of interest (MOI) plot:

```{r}
MOI_plot(met,
         group_factor="tumor_normal",
         MOI = 2,
         label_colors=c("darkseagreen","dodgerblue"))
```

The full case study of the example data set is part of the package vignette "Metabolic correlation network analysis".

# Session information {.unnumbered}

```{r}
sessionInfo()
```

[^1]: Priolo, C., Pyne, S., Rose, J., Regan, E. R., Zadra, G., Photopoulos, C., et al. (2014). AKT1 and MYC Induce Distinctive Metabolic Fingerprints in Human Prostate Cancer. Cancer Research, 74(24), 7198–7204. http://doi.org/10.1158/0008-5472.CAN-14-1490

[^2]: Sig M (2017). MultiAssayExperiment: Software for the integration of multi-omics experiments in Bioconductor. R package version 1.2.1

[^3]: Jewison T, Su Y, Disfany FM, et al. SMPDB 2.0: Big Improvements to the Small Molecule Pathway Database. Nucleic Acids Res. 2014 Jan;42(Database issue):D478-84.

[^4]: Armitage, E. G., Godzien, J., Alonso-Herranz, V., L pez-Gonz lvez, N., & Barbas, C. (2015). Missing value imputation strategies for metabolomics data. Electrophoresis, 36(24), 3050–3060. http://doi.org/10.1002/elps.201500352

[^5]: Huber, W., Heydebreck, von, A., Sültmann, H., Poustka, A., & Vingron, M. (2002). Variance stabilization applied to microarray data calibration and to the quantification of differential expression. Bioinformatics, 18 Suppl 1, S96–104.

[^6]: Ignatiadis, N., Klaus, B., Zaugg, J. B., & Huber, W. (2016). Data-driven hypothesis weighting increases detection power in genome-scale multiple testing. Nature Methods, 13(7), 577–580. http://doi.org/10.1038/nmeth.3885

[^7]: Langfelder, P., & Horvath, S. (2008). WGCNA: an R package for weighted correlation network analysis. BMC Bioinformatics, 9, 559–559. http://doi.org/10.1186/1471-2105-9-559

[^8]: Zheng, C.-H., Yuan, L., Sha, W., & Sun, Z.-L. (2014). Gene differential coexpression analysis based on biweight correlation and maximum clique. BMC Bioinformatics, 15 Suppl 15(Suppl 15), S3. http://doi.org/10.1186/1471-2105-15-S15-S3

[^9]: Zhang, B., & Horvath, S. (2005). A general framework for weighted gene co-expression network analysis. Statistical Applications in Genetics and Molecular Biology, 4(1), Article17. http://doi.org/10.2202/1544-6115.1128

[^10]: Langfelder, P., Zhang, B., & Horvath, S. (2008). Defining clusters from a hierarchical cluster tree: the Dynamic Tree Cut package for R. Bioinformatics, 24(5), 719–720. http://doi.org/10.1093/bioinformatics/btm563.

[^11]: Horvath, S., & Dong, J. (2008). Geometric Interpretation of Gene Coexpression Network Analysis. PLoS Computational Biology (PLOSCB) 4(8), 4(8), e1000117–e1000117. http://doi.org/10.1371/journal.pcbi.1000117
